generate-code.sh

```diff
@@ -128,7 +138,49 @@ _generateData() {
       "${proto_src}"
 }
 
+_generateValidationTests() {
+    # Source the schema protos.
+    proto_dir="${GENERATE_DATA_SOURCE}/proto"
+    proto_src=$(realpath --relative-to="${proto_dir}" "$1")
+    # Derive path to data.proto from event.proto.
+    data_src=$(dirname "${proto_src}")/data.proto
+    # Derive destination directory.
+    code_dest=$(dirname "$(dirname "${proto_src}")" | cut -d'/' -f3-)
+    # Product name & API version
+    product=$(basename "$code_dest")
+    version=$(basename "$(dirname "${proto_src}")")
+
+    # Explicit type versioning after v1.
+    if [[ "${version}" == "v1" ]]; then
+      version=""
+    fi
+
+    # Skip unsupported products.
+    if [[ "${product}" == "pubsub" ]]; then
+      echo "- ${product}: skipping generation, not currently supported"
+      return
+    fi
+
+    echo "- ${product}: ${proto_src} => ${code_dest}data${version}/data_test.go"
+
+    $GENERATE_PROTOC_PATH --go-googlecetypes_out=. \
+        --go-googlecetypes_opt="Mgoogle/events/cloudevent.proto"="github.com/googleapis/google-cloudevents-go/thirdparty/cloudevents;cloudevents" \
+        --go-googlecetypes_opt="M${proto_src}"="${code_dest}data${version}" \
+        --go-googlecetypes_opt="M${data_src}"="${code_dest}data${version};${product}data${version}" \
```

Comment:
Is this for redirecting the generated outputs?
If so, the `--paths=source_relative` flag might be a cleaner method (https://developers.google.com/protocol-buffers/docs/reference/go-generated#invocation).
combining it with `module` might even let us remove some of the path munging above.

Reply:
This is being used to map protos/google/events/cloud/auditlog/v1 => cloud/auditlogdata.

While I could remove this from the shell script and do it in the protogen Go code, I kept this in bash because it parallels the use of protoc-gen-go above, where the source_relative approach does not work unless we start injecting `go_package` option into the protos via copybara to replicate this path mapping.

Reply:
I'm not proposing changing the generator Go code. I'm pointing out that there *might* be a simpler set of arguments to `protoc` that can provide the same functionality, in a more robust way than `dirname` and `cut`. I believe this is what the `module` option is for, and it is useful without adding `go_package`.

My understanding, using your example, is that with the file `protos/google/events/cloud/auditlog/v1/data.proto`, and a `module=${whatever_the_inferred_go_package_is}/protos/google/events`, we'd get an output file in `cloud/auditlog/v1/data.pb2.go` - This is slightly different from what you're doing here, but if it works, I believe it is simpler and more maintainable.

A similar approach could be applied to the prior generator if you value consistency, though I do not have a strong opinion there.

Reply:
From my research this doesn't work as expected without proto changes. Discussed and pushing this out of current PR scope.